feat(dap): send Disconnect if Terminated event received

filipdutescu opened the pull request:
Send a `Disconnect` DAP request if the `Terminated` event is received. According to the specification, if the debugging session was started by as `launch`, the debuggee should be terminated alongside the session. If instead the session was started as `attach`, it should not be disposed of.

This default behaviour can be overriden if the `supportTerminateDebuggee` capability is supported by the adapter, through the `Disconnect` request `terminateDebuggee` argument, as described in
[the specification][discon-spec].

This also implies saving the starting command for a debug sessions, in order to decide which behaviour should be used, as well as validating the capabilities of the adapter, in order to decide what the disconnect should do.

An additional change made is handling of the `Exited` event, showing a message if the exit code is different than `0`, for the user to be aware off the termination failure.

[discon-spec]: https://microsoft.github.io/debug-adapter-protocol/specification#Requests_Disconnect

Closes: #4674
Signed-off-by: Filip Dutescu <filip.dutescu@gmail.com>

filipdutescu commented on helix-view/src/handlers/dap.rs:
```diff
@@ -274,6 +275,62 @@ impl Editor {
+                    let starting_request = debugger.starting_request();
+                    let result = debugger.disconnect(disconnect_args).await;
+                    if result.is_ok() {
+                        if restart {
+                            if let StartingRequest::None = starting_request {
+                                self.set_error("No starting request found, to be used in restarting the debugging session.");
+                            } else {
+                                log::info!("Attempting to restart debug session.");
+                                let restart_args = restart_args.unwrap();
+                                let relaunch_resp =
+                                    if let StartingRequest::Launch = starting_request {
+                                        debugger.launch(restart_args).await
+                                    } else {
+                                        debugger.attach(restart_args).await
+                                    };
+
+                                if let Err(err) = relaunch_resp {
+                                    self.set_error(format!(
+                                        "Failed to restart debugging session: {:?}",
+                                        err
+                                    ));
+                                }
+                            }
```
restart should be done with the [`Restart Request`](https://microsoft.github.io/debug-adapter-protocol/specification#Requests_Restart) :thinking:

filipdutescu replied:
I was incorrect in my assumption. the request should be used to _*trigger*_ a restart, rather than "create one". In any case, this should be further elaborated, most likely in a different PR, handling restarting as a whole

filipdutescu reviewed (commented):
just realised the restart implementation I made is not correct according to the spec. will redo it

filipdutescu commented:
I have re-read the spec and I believe this part of the restart sequence to be correct. What needs to be done in addition is to expose a restart request, if supported, to the user, same as with launch, terminate etc.

Cc: @pascalkuthe could we move it to waiting on review? Many thanks!

archseer commented on helix-dap/src/client.rs:
```diff
@@ -41,6 +43,13 @@ pub struct Client {
+#[derive(Clone, Copy, Debug)]
+pub enum StartingRequest {
+    Launch,
+    Attach,
```
Remove `None` and instead use `Option<StartingRequest>`

archseer replied:
Also this seems better named as connection type?

archseer commented on helix-dap/src/client.rs:
```diff
@@ -78,7 +86,7 @@ impl Client {
             request_counter: AtomicU64::new(0),
             caps: None,
-            //
+            starting_request: None,
```
This should also be renamed to connection_type now

archseer commented on helix-dap/src/client.rs:
```diff
@@ -207,6 +215,10 @@ impl Client {
         self.id
     }
 
+    pub fn starting_request(&self) -> Option<ConnectionType> {
```
Same here

archseer commented on helix-view/src/handlers/dap.rs:
```diff
@@ -274,6 +275,65 @@ impl Editor {
+                    if result.is_ok() {
+                        if restart {
+                            match starting_request {
+                                None => {
```
None should come after Some

archseer commented on helix-view/src/handlers/dap.rs:
```diff
@@ -274,6 +275,65 @@ impl Editor {
+
+                    let starting_request = debugger.starting_request();
+                    let result = debugger.disconnect(disconnect_args).await;
+                    if result.is_ok() {
```
Rewrite this so it's 

```
if let Err(ere) = result {
   set_error(...)
   return;
}
```

That way the error exits early and the regular path isn't indented inside an if

archseer commented on helix-view/src/handlers/dap.rs:
```diff
@@ -274,6 +275,65 @@ impl Editor {
+                    let starting_request = debugger.starting_request();
+                    let result = debugger.disconnect(disconnect_args).await;
+                    if result.is_ok() {
+                        if restart {
```
Similar to the other PR you want to match on the argument so we avoid an unwrap

filipdutescu commented:
@archseer implemented your comments, thanks very much for the patience and advice, I have learnt a lot and hopefully improved the way I write Rust. I also rebased the PR and squashed the commits, to make it a bit cleaner, since there are not that many modifications (hopefully) as to make it a chore to revisit the changes.

archseer reviewed (approved):
Sorry it took me a while to get back to this! LGTM!

filipdutescu commented:
No worries, thanks for reviewing and merging it!
